<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Aligning single cell RNA to histology tissue images • gcproc</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Aligning single cell RNA to histology tissue images" />
<meta property="og:description" content="gcproc" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gcproc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Aligning_scRNAseq_to_Spatial_Transcriptomics.html">Aligning single cell RNA to histology tissue images</a>
    </li>
    <li>
      <a href="../articles/Dimensionality_reduction_with_gcproc.html">Dimensionality reduction with gcproc</a>
    </li>
    <li>
      <a href="../articles/Neurips_OpenSingleCell_Prediction_task.html">Neurips Open Problems in Single Cell Analysis: Prediction task</a>
    </li>
    <li>
      <a href="../articles/Recovering_missing_test_data_points.html">Recovering missing test data points</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<script src="Aligning_scRNAseq_to_Spatial_Transcriptomics_files/header-attrs-2.11/header-attrs.js"></script>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Aligning single cell RNA to histology tissue images</h1>
            
            <h4 class="date">Compiled: October 07, 2021</h4>
      
      
      <div class="hidden name"><code>Aligning_scRNAseq_to_Spatial_Transcriptomics.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level3">
<h3>Introduction</h3>
<p>Generalised Canonical Procrustes (gcproc) can align datasets of different modalities with varying data structure (e.g. sample by feature matrix, pixel image data, …) and varying dimensions. It can encode and transform image data to extract relevant features to be analysed jointly with other modalities. With parameters that encode multiple datasets into the same space, the weights act as learned functions that can (inverse) transform from one modality to another.</p>
<p>Here, a novel dataset from a variety of experimental methods called Spatial Transcriptomic Technology (example used here is thanks to NanoString’s Hackathon @ DevPost) that combines histology imaging tissue with measured gene expression, can be used to:</p>
<p>1 - Encode the histology pixel data into a representative space for feature extraction</p>
<p>2 - Recover gene expression using histology tissue with a few gene biomarkers relevant to the tissue</p>
</div>
<div id="step-0---prepare-libraries" class="section level3">
<h3>Step 0 - Prepare libraries</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gcproc)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mclust)</span></code></pre></div>
<pre><code>## Package &#39;mclust&#39; version 5.4.7
## Type &#39;citation(&quot;mclust&quot;)&#39; for citing this R package in publications.</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(irlba)</span></code></pre></div>
<pre><code>## Loading required package: Matrix</code></pre>
</div>
<div id="step-1---prepare-datasets-from-nanostring-geomx-spatial-transcriptomic-technology" class="section level3">
<h3>Step 1 - Prepare datasets from NanoString GeoMX Spatial Transcriptomic Technology</h3>
<p>Kidney spatial data is gratefully taken from the NanoString Hackathon @ DevPost. Here, 231 regions of interest have had gene expression measured at a spatial resolution. Corresponding overlapping image information of immunofluorescent kidney tissue is converted to a pixel by tile matrix, where each tile is 32 by 32 pixels by 3 Red-Green-Blue channels.</p>
<p>Single cell kidney data is also taken from <a href="https://pubmed.ncbi.nlm.nih.gov/29980650/">Wu et al (2018)</a> to be aligned to immunofluorescent kidney tissue. Here, the top 2000 shared variably genes are taken for all datasets containing gene expression.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>kidney_nanostring_rnaseq_train <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;../../data/kidney_nanostring_spatial/Kidney_Raw_TargetCountMatrix.txt&quot;</span>,<span class="at">row.names=</span><span class="dv">1</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>kidney_nanostring_pixel_train <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;../../data/kidney_nanostring_spatial/TRAIN_pixels_color.csv&quot;</span>,<span class="at">row.names=</span><span class="dv">1</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>kidney_sc_rnaseq <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;../../data/kidney_nanostring_spatial/wu_counts.csv&quot;</span>,<span class="at">row.names=</span><span class="dv">1</span>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>matched_genes <span class="ot">&lt;-</span> <span class="fu">row.names</span>(kidney_nanostring_rnaseq_train)[<span class="fu">row.names</span>(kidney_nanostring_rnaseq_train) <span class="sc">%in%</span> <span class="fu">row.names</span>(kidney_sc_rnaseq)]</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>top_shared_var_genes <span class="ot">&lt;-</span> <span class="fu">sort</span>(matched_genes[<span class="fu">order</span>(<span class="fu">apply</span>(kidney_nanostring_rnaseq_train[matched_genes,],<span class="dv">1</span>,var)<span class="sc">+</span><span class="fu">apply</span>(kidney_sc_rnaseq[matched_genes,],<span class="dv">1</span>,var),<span class="at">decreasing =</span> T)[<span class="dv">1</span><span class="sc">:</span><span class="dv">1000</span>]])</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>kidney_nanostring_rnaseq_train.select <span class="ot">&lt;-</span> <span class="fu">t</span>(kidney_nanostring_rnaseq_train[top_shared_var_genes,])</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>kidney_sc_rnaseq.select <span class="ot">&lt;-</span> <span class="fu">t</span>(kidney_sc_rnaseq[top_shared_var_genes,])</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>kidney_nanostring_rnaseq_train.select.copy <span class="ot">&lt;-</span> kidney_nanostring_rnaseq_train.select <span class="ot">&lt;-</span> kidney_nanostring_rnaseq_train.select</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>kidney_nanostring_pixel_train.select <span class="ot">&lt;-</span> kidney_nanostring_pixel_train</span></code></pre></div>
</div>
<div id="step-2---initialise-gcproc-parameters-for-both-pixel-encoding-and-gene-recovery" class="section level3">
<h3>Step 2 - Initialise gcproc parameters for both pixel encoding and gene recovery</h3>
<p>The default parameters in the configuration variable are modified to adjust for the data.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>config <span class="ot">&lt;-</span> gcproc<span class="sc">::</span><span class="fu">extract_config</span>(<span class="at">verbose=</span>F)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>config<span class="sc">$</span>i_dim <span class="ot">&lt;-</span> <span class="dv">50</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>config<span class="sc">$</span>j_dim <span class="ot">&lt;-</span> <span class="dv">50</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>config<span class="sc">$</span>max_iter <span class="ot">&lt;-</span> <span class="dv">30</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>config<span class="sc">$</span>verbose <span class="ot">&lt;-</span> F</span></code></pre></div>
</div>
<div id="step-3---run-gcproc-on-histology-immunofluorescent-images-to-extract-features-from-pixel-data" class="section level3">
<h3>Step 3 - Run gcproc on histology immunofluorescent images to extract features from pixel data</h3>
<p>Spatial Transcriptomic regions of interest (with gene expression measurements) of immunofluorescent kidney histology images are converted to tabular form by cutting the image into tiles of 32 by 32 pixels, and taking each tile as a sample, with pixels as the features. To extract relevant features that are not based on highly variable image noise and perturbations (e.g. rotation, pixel intensity, …), the pixel dimension is encoded into a reduced subspace to give information of higher quality.</p>
<p>The top 21 components out of 70 are plotted to reveal the structure and profile of the tile, each representing the spatial tissue and cell morphology.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>data_list.encode <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">y=</span><span class="fu">as.matrix</span>(kidney_nanostring_pixel_train.select),</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">as.matrix</span>(kidney_sc_rnaseq)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>gcproc.encode <span class="ot">&lt;-</span> gcproc<span class="sc">::</span><span class="fu">gcproc</span>(</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_list =</span> data_list.encode,</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">config =</span> config</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(png)</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(grid)</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gridExtra)</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>convert_to_HEAT <span class="ot">&lt;-</span> <span class="cf">function</span>(x){</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>  e <span class="ot">&lt;-</span> <span class="fu">ecdf</span>(x)</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>  j <span class="ot">&lt;-</span> <span class="fu">e</span>(x)</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">array</span>(j,<span class="at">dim=</span><span class="fu">dim</span>(x))</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(x)</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot top 21 components</span></span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>gl <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">21</span>),<span class="cf">function</span>(i){</span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a>  g <span class="ot">&lt;-</span> (<span class="fu">convert_to_HEAT</span>(<span class="fu">aperm</span>(<span class="fu">array</span>(gcproc.encode<span class="sc">$</span>dimension_reduction[[<span class="dv">1</span>]]<span class="sc">$</span>feature_x.dim_reduce.code[,i],<span class="at">dim=</span><span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">32</span>,<span class="dv">32</span>)),<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">1</span>))))</span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a>  g <span class="ot">&lt;-</span> <span class="fu">rasterGrob</span>(g, <span class="at">interpolate=</span><span class="cn">TRUE</span>)</span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a>  p.main <span class="ot">&lt;-</span> ggplot2<span class="sc">::</span><span class="fu">qplot</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">geom=</span><span class="st">&quot;blank&quot;</span>) <span class="sc">+</span></span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">annotation_custom</span>(g, <span class="at">xmin=</span><span class="sc">-</span><span class="cn">Inf</span>, <span class="at">xmax=</span><span class="cn">Inf</span>, <span class="at">ymin=</span><span class="sc">-</span><span class="cn">Inf</span>, <span class="at">ymax=</span><span class="cn">Inf</span>)</span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a>  p.main</span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a>gl <span class="ot">&lt;-</span> <span class="fu">lapply</span>(gl,<span class="cf">function</span>(X){</span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a>    X <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.line=</span><span class="fu">element_blank</span>(),<span class="at">axis.text.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a>              <span class="at">axis.text.y=</span><span class="fu">element_blank</span>(),<span class="at">axis.ticks=</span><span class="fu">element_blank</span>(),</span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a>              <span class="at">axis.title.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a>              <span class="at">axis.title.y=</span><span class="fu">element_blank</span>(),<span class="at">legend.position=</span><span class="st">&quot;none&quot;</span>,</span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true" tabindex="-1"></a>              <span class="at">panel.background=</span><span class="fu">element_blank</span>(),<span class="at">panel.border=</span><span class="fu">element_blank</span>(),<span class="at">panel.grid.major=</span><span class="fu">element_blank</span>(),</span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true" tabindex="-1"></a>              <span class="at">panel.grid.minor=</span><span class="fu">element_blank</span>(),<span class="at">plot.background=</span><span class="fu">element_blank</span>())</span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-47"><a href="#cb7-47" aria-hidden="true" tabindex="-1"></a>lm <span class="ot">&lt;-</span> <span class="fu">rbind</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">6</span>,<span class="dv">7</span>),</span>
<span id="cb7-48"><a href="#cb7-48" aria-hidden="true" tabindex="-1"></a>            <span class="fu">c</span>(<span class="dv">8</span>,<span class="dv">9</span>,<span class="dv">10</span>,<span class="dv">11</span>,<span class="dv">12</span>,<span class="dv">13</span>,<span class="dv">14</span>),</span>
<span id="cb7-49"><a href="#cb7-49" aria-hidden="true" tabindex="-1"></a>            <span class="fu">c</span>(<span class="dv">15</span>,<span class="dv">16</span>,<span class="dv">17</span>,<span class="dv">18</span>,<span class="dv">19</span>,<span class="dv">20</span>,<span class="dv">21</span>)</span>
<span id="cb7-50"><a href="#cb7-50" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb7-51"><a href="#cb7-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-52"><a href="#cb7-52" aria-hidden="true" tabindex="-1"></a>gg1 <span class="ot">&lt;-</span> <span class="fu">arrangeGrob</span>(</span>
<span id="cb7-53"><a href="#cb7-53" aria-hidden="true" tabindex="-1"></a>  <span class="at">grobs =</span> gl,</span>
<span id="cb7-54"><a href="#cb7-54" aria-hidden="true" tabindex="-1"></a>  <span class="at">layout_matrix =</span> lm</span>
<span id="cb7-55"><a href="#cb7-55" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb7-56"><a href="#cb7-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-57"><a href="#cb7-57" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(gg1)</span></code></pre></div>
<p><img src="Aligning_scRNAseq_to_Spatial_Transcriptomics_files/figure-html/step_3-1.png" width="960" /></p>
</div>
<div id="step-4---recover-gene-expression-on-histology-tissue-with-a-few-gene-markers-present" class="section level3">
<h3>Step 4 - Recover gene expression on histology tissue with a few gene markers present</h3>
<p>With the pixel information encoded to better represent the immunofluorescent kidney tissue, the newly encoded features can be used for further analysis. The transfer of parameters can be done via the anchor framework. Just as an anchor would position a ship at sea, an anchored parameter would freeze the parameter in place, while learning other parameters jointly in relation to the anchor.</p>
<p>An example analysis is the recovery of missing gene expression, imputed from image-based histology with a few kidney gene markers.</p>
<p>Here, 201 regions of interest (with pixel and all gene data) are randomly sampled from the total of 231 regions and treated as the training set. Thus, 30 regions are taken as missing and treated as the test set (with only pixel data and a list of key gene kidney markers). The key kidney gene markers assumed present in both training and test sets are PODXL, WT1, PAX2, VIM, AQP1, NPHS1, and B2M all known to be present in kidney cells.</p>
<p>The aim is to recover all other non-key genes via the pixels within the 30 regions in the test set, while taking the 201 other regions as the training set to be the main modelling source.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>gene_marker_present_in_histology_taken_as_example <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;PODXL&quot;</span>,<span class="st">&quot;WT1&quot;</span>,<span class="st">&quot;PAX2&quot;</span>,<span class="st">&quot;VIM&quot;</span>,<span class="st">&quot;AQP1&quot;</span>,<span class="st">&quot;NPHS1&quot;</span>,<span class="st">&quot;B2M&quot;</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">10</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>train_ids <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">231</span>),<span class="dv">201</span>)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>design_recover <span class="ot">&lt;-</span> gcproc<span class="sc">::</span><span class="fu">extract_recovery_framework</span>(F)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>design_recover.x <span class="ot">&lt;-</span> kidney_nanostring_rnaseq_train.select <span class="sc">-</span> kidney_nanostring_rnaseq_train.select</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>design_recover.x[<span class="sc">-</span>train_ids,<span class="fu">which</span>(<span class="sc">!</span><span class="fu">colnames</span>(kidney_nanostring_rnaseq_train.select) <span class="sc">%in%</span> <span class="fu">c</span>(gene_marker_present_in_histology_taken_as_example))] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>design_recover.x[train_ids,<span class="fu">which</span>(<span class="fu">colnames</span>(kidney_nanostring_rnaseq_train.select) <span class="sc">%in%</span> <span class="fu">c</span>(gene_marker_present_in_histology_taken_as_example))] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>design_recover<span class="sc">$</span>method <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;knn&quot;</span>)</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>design_recover<span class="sc">$</span>design.list <span class="ot">&lt;-</span> <span class="fu">list</span>( <span class="at">y=</span><span class="cn">NULL</span>,</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">x=</span><span class="fu">as.matrix</span>(design_recover.x)</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>kidney_nanostring_rnaseq_train.select <span class="ot">&lt;-</span> kidney_nanostring_rnaseq_train.select.copy<span class="sc">*</span>(<span class="dv">1</span><span class="sc">-</span>design_recover.x) <span class="sc">+</span> <span class="fu">do.call</span>(<span class="st">&#39;cbind&#39;</span>,<span class="fu">lapply</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">dim</span>(design_recover.x)[<span class="dv">2</span>]),<span class="cf">function</span>(X){(design_recover.x[,X])<span class="sc">*</span><span class="fu">median</span>(kidney_nanostring_rnaseq_train.select.copy[<span class="fu">which</span>((<span class="dv">1</span><span class="sc">-</span>design_recover.x)[,X]<span class="sc">==</span><span class="dv">1</span>),X])}))</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>data_list.transfer <span class="ot">&lt;-</span> <span class="fu">list</span>( <span class="at">y=</span><span class="fu">as.matrix</span>(kidney_nanostring_pixel_train.select),</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">x=</span><span class="fu">as.matrix</span>(kidney_nanostring_rnaseq_train.select)</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>join <span class="ot">&lt;-</span> gcproc<span class="sc">::</span><span class="fu">extract_join_framework</span>(F)</span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>join<span class="sc">$</span>alpha <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>)</span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gcproc)</span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>gcproc.transfer <span class="ot">&lt;-</span> gcproc<span class="sc">::</span><span class="fu">gcproc</span>(</span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>               <span class="at">data_list =</span> data_list.transfer,</span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a>               <span class="at">config =</span> config,</span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a>               <span class="at">recover =</span> design_recover, </span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a>               <span class="at">join =</span> join</span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a>               )</span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a>corr_genes <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">&#39;c&#39;</span>,<span class="fu">lapply</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">dim</span>(kidney_nanostring_rnaseq_train.select.copy)[<span class="dv">2</span>]),<span class="cf">function</span>(i){</span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cor</span>(kidney_nanostring_rnaseq_train.select.copy[<span class="sc">-</span>train_ids,i],(gcproc.transfer<span class="sc">$</span>recover<span class="sc">$</span>predict.list[[<span class="dv">2</span>]])[<span class="sc">-</span>train_ids,i])</span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a>}))</span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a>other_than_key_gene.id <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="sc">!</span><span class="fu">colnames</span>(kidney_nanostring_rnaseq_train.select) <span class="sc">%in%</span> <span class="fu">c</span>(gene_marker_present_in_histology_taken_as_example))</span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a>corr_cells_remove_key_genes <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">&#39;c&#39;</span>,<span class="fu">lapply</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">dim</span>(kidney_nanostring_rnaseq_train.select.copy)[<span class="dv">1</span>])[<span class="sc">-</span>train_ids],<span class="cf">function</span>(i){</span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cor</span>(kidney_nanostring_rnaseq_train.select.copy[i,other_than_key_gene.id],(gcproc.transfer<span class="sc">$</span>recover<span class="sc">$</span>predict.list[[<span class="dv">2</span>]])[i,other_than_key_gene.id])</span>
<span id="cb8-41"><a href="#cb8-41" aria-hidden="true" tabindex="-1"></a>}))</span>
<span id="cb8-42"><a href="#cb8-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-43"><a href="#cb8-43" aria-hidden="true" tabindex="-1"></a>corr_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Axis =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;Gene Correlations&quot;</span>,<span class="fu">length</span>(corr_genes)),<span class="fu">rep</span>(<span class="st">&quot;Cell Correlations&quot;</span>,<span class="fu">length</span>(corr_cells_remove_key_genes))), <span class="at">Pearson =</span> <span class="fu">c</span>(corr_genes,corr_cells_remove_key_genes))</span>
<span id="cb8-44"><a href="#cb8-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-45"><a href="#cb8-45" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb8-46"><a href="#cb8-46" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(corr_data, <span class="fu">aes</span>(<span class="at">x=</span>Axis, <span class="at">y=</span>Pearson)) <span class="sc">+</span> </span>
<span id="cb8-47"><a href="#cb8-47" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>()</span></code></pre></div>
<p><img src="Aligning_scRNAseq_to_Spatial_Transcriptomics_files/figure-html/step_4-1.png" width="768" /></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfcol=</span><span class="fu">c</span>(<span class="dv">4</span>,<span class="dv">2</span>))</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">which</span>(<span class="fu">colnames</span>(kidney_nanostring_rnaseq_train.select.copy)<span class="sc">%in%</span><span class="fu">c</span>(<span class="st">&quot;HLA-A&quot;</span>,<span class="st">&quot;CD74&quot;</span>,<span class="st">&quot;AQP2&quot;</span>,<span class="st">&quot;KRT8&quot;</span>,<span class="st">&quot;NPHS2&quot;</span>,<span class="st">&quot;ITGB1&quot;</span>,<span class="st">&quot;PAX8&quot;</span>,<span class="st">&quot;PDGFRA&quot;</span>))){</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  adj.r.squared <span class="ot">&lt;-</span> <span class="fu">summary</span>(<span class="fu">lm</span>(kidney_nanostring_rnaseq_train.select.copy[<span class="sc">-</span>train_ids,i]<span class="sc">~</span>(gcproc.transfer<span class="sc">$</span>recover<span class="sc">$</span>predict.list[[<span class="dv">2</span>]])[<span class="sc">-</span>train_ids,i]))<span class="sc">$</span>adj.r.squared</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(<span class="at">x =</span> (gcproc.transfer<span class="sc">$</span>recover<span class="sc">$</span>predict.list[[<span class="dv">2</span>]])[<span class="sc">-</span>train_ids,i],<span class="at">y =</span> kidney_nanostring_rnaseq_train.select.copy[<span class="sc">-</span>train_ids,i],<span class="at">main=</span><span class="fu">paste</span>(<span class="fu">colnames</span>(kidney_nanostring_rnaseq_train.select)[i],<span class="st">&quot; </span><span class="sc">\n</span><span class="st"> Adj. R^2: &quot;</span>,adj.r.squared,<span class="at">sep=</span><span class="st">&quot;&quot;</span>), <span class="at">xlab =</span> <span class="st">&quot;recovered RNA levels&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Observed RNA levels&quot;</span>,<span class="at">cex=</span><span class="fl">1.5</span>,<span class="at">pch=</span><span class="dv">19</span>)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p><img src="Aligning_scRNAseq_to_Spatial_Transcriptomics_files/figure-html/step_4-2.png" width="768" /></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer>
      <div class="copyright">
  <p>Developed by First Last.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>
